File: API/components/clients.ts
import { players } from "../api";
import { ServerMessage } from "../data/server_message";
import { Client } from "./client";
import { WebSocket } from "ws"
import { v4 as uuidv4 } from 'uuid'


class Clients {
    private readonly clients: Client[] = []

    findFromSocket = (socket: WebSocket): Client => {
        return this.clients.find((c: Client) => {
            return c.tcp?.socket == socket
        })
    }

    findFromAddressAndPort = (address: string, port: number): Client => {
        return this.clients.find((c: Client) => {
            return c.udp?.port == port && c.udp?.address == address
        })
    }

    findFromId = (id: string): Client => {
        return this.clients.find((c: Client) => c.id == id)
    }

    create = (): Client => {
        const id: string = uuidv4()
        const newClient: Client = new Client(id)
        this.clients.push(newClient)
        console.log("[CLIENTS] " + this.clients.map(c => c.id))
        return newClient
    }

    sendToAllPlayersUDP = (udpMessage: ServerMessage): void => {
        this.clients.forEach((c: Client) => {
            if (c.udp != null && players.findFromId(c.id) != null)
                c.udp.send(udpMessage)
        })
    }

    sendToAllPlayersTCP = (tcpMessage: ServerMessage): void => {
        this.clients.forEach((c: Client) => {
            if (c.udp != null && players.findFromId(c.id) != null)
                c.tcp.send(tcpMessage)
        })
    }
}

export { Clients }

File: API/TestUDP/src/clients/Clients.cs
namespace TestUDP;

public class Clients
{
    private readonly List<Client> _clients = new List<Client>();

    public Client Create(TCPClient tcpClient) {
        Client newClient = new Client(tcpClient);
        _clients.Add(newClient);
        Console.WriteLine($"[Clients] {_clients.Count}");
        return newClient;
    }

    public void Remove(Client client) {
        _clients.Remove(client);
        Console.WriteLine($"[Clients] {_clients.Count}");
    }

    public void Remove(TCPClient tcpClient) {
        int index = _clients.FindIndex((Client c) => { return c.Tcp == tcpClient; });
        if (index != -1)
            _clients.RemoveAt(index);
        Console.WriteLine($"[Clients] {_clients.Count}");
    }

    public Client? Find(string address, int port) {
        return _clients.Find((Client c) => c.Udp?.Address == address && c.Udp.Port == port);
    }

    public Client? Find(TCPClient tcpClient) {
        return _clients.Find((Client c) => c.Tcp == tcpClient);
    }

    //public Client Find(string id) {
    //    return _clients.Find((Client c) => c.Id == id);
    //}

    public Client? Find(string secret) {
        return _clients.Find((Client c) => c.Secret == secret);
    }
}
